### ISSUE TITLE
UnboundLocalError: local variable 'master_display_' referenced before assignment

### ISSUE THREAD

**Issue by faruktufekci** [pycaret/pycaret#788]
Bug? (similar to visualizer --> viz issue before)

```
---------------------------------------------------------------------------
UnboundLocalError                         Traceback (most recent call last)
<ipython-input-20-9ff5ccfe3077> in <module>
----> 1 compare_models()

~/caret/lib/python3.7/site-packages/pycaret/classification.py in compare_models(include, exclude, fold, round, cross_validation, sort, n_select, budget_time, turbo, errors, fit_kwargs, groups, verbose)
    775         fit_kwargs=fit_kwargs,
    776         groups=groups,
--> 777         verbose=verbose,
    778     )
    779 

~/caret/lib/python3.7/site-packages/pycaret/internal/tabular.py in compare_models(include, exclude, fold, round, cross_validation, sort, n_select, budget_time, turbo, errors, fit_kwargs, groups, verbose, display)
   2219 
   2220     compare_models_ = (
-> 2221         master_display_.apply(
   2222             highlight_max,
   2223             subset=[

UnboundLocalError: local variable 'master_display_' referenced before assignment
```

**Comment by Yard1** [comment]
Can you send us your `logs.log` file, present in the directory from where you run the script/notebook?

**Comment by pycaret** [comment]
@faruktufekci I can't reproduce this error. If you can please share the logs file as @Yard1 requested, also if you can upload the Notebook here? That will help us troubleshoot.

**Comment by garar** [comment]
Hello!

I can't paste logs here, because it contains confidential data, but I can say that this when an exception is thrown when creating model for comparison.

In my case, this only happens I use custom metrics, example:

```python
add_metric('logloss', 'LogLoss', log_loss, greater_is_better=False)
```

Then I call:

```python
top_models = compare_models(n_select=5)
```

I see this exception logs.log

```
2020-10-30 22:56:13,038:ERROR:joblib.externals.loky.process_executor._RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/home/garar/.local/share/virtualenvs/pycaret-ml-W1zZIU60/lib/python3.8/site-packages/joblib/externals/loky/process_executor.py", line 431, in _process_worker
    r = call_item()
  File "/home/garar/.local/share/virtualenvs/pycaret-ml-W1zZIU60/lib/python3.8/site-packages/joblib/externals/loky/process_executor.py", line 285, in __call__
    return self.fn(*self.args, **self.kwargs)
  File "/home/garar/.local/share/virtualenvs/pycaret-ml-W1zZIU60/lib/python3.8/site-packages/joblib/_parallel_backends.py", line 595, in __call__
    return self.func(*args, **kwargs)
  File "/home/garar/.local/share/virtualenvs/pycaret-ml-W1zZIU60/lib/python3.8/site-packages/joblib/parallel.py", line 262, in __call__
    return [func(*args, **kwargs)
  File "/home/garar/.local/share/virtualenvs/pycaret-ml-W1zZIU60/lib/python3.8/site-packages/joblib/parallel.py", line 262, in <listcomp>
    return [func(*args, **kwargs)
  File "/home/garar/.local/share/virtualenvs/pycaret-ml-W1zZIU60/lib/python3.8/site-packages/sklearn/model_selection/_validation.py", line 560, in _fit_and_score
    test_scores = _score(estimator, X_test, y_test, scorer)
  File "/home/garar/.local/share/virtualenvs/pycaret-ml-W1zZIU60/lib/python3.8/site-packages/sklearn/model_selection/_validation.py", line 607, in _score
    scores = scorer(estimator, X_test, y_test)
  File "/home/garar/.local/share/virtualenvs/pycaret-ml-W1zZIU60/lib/python3.8/site-packages/sklearn/metrics/_scorer.py", line 87, in __call__
    score = scorer._score(cached_call, estimator,
  File "/home/garar/.local/share/virtualenvs/pycaret-ml-W1zZIU60/lib/python3.8/site-packages/sklearn/metrics/_scorer.py", line 212, in _score
    return self._sign * self._score_func(y_true, y_pred,
  File "/home/garar/.local/share/virtualenvs/pycaret-ml-W1zZIU60/lib/python3.8/site-packages/sklearn/utils/validation.py", line 72, in inner_f
    return f(**kwargs)
  File "/home/garar/.local/share/virtualenvs/pycaret-ml-W1zZIU60/lib/python3.8/site-packages/sklearn/metrics/_classification.py", line 2226, in log_loss
    raise ValueError("y_true and y_pred contain different number of "
ValueError: y_true and y_pred contain different number of classes 3, 2. Please provide the true labels explicitly through the labels argument. Classes found in y_true: [0 1 2]
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/garar/.local/share/virtualenvs/pycaret-ml-W1zZIU60/lib/python3.8/site-packages/pycaret/internal/tabular.py", line 2149, in compare_models
    model, model_fit_time = create_model_supervised(
  File "/home/garar/.local/share/virtualenvs/pycaret-ml-W1zZIU60/lib/python3.8/site-packages/pycaret/internal/tabular.py", line 3058, in create_model_supervised
    scores = cross_validate(
  File "/home/garar/.local/share/virtualenvs/pycaret-ml-W1zZIU60/lib/python3.8/site-packages/sklearn/utils/validation.py", line 72, in inner_f
    return f(**kwargs)
  File "/home/garar/.local/share/virtualenvs/pycaret-ml-W1zZIU60/lib/python3.8/site-packages/sklearn/model_selection/_validation.py", line 242, in cross_validate
    scores = parallel(
  File "/home/garar/.local/share/virtualenvs/pycaret-ml-W1zZIU60/lib/python3.8/site-packages/joblib/parallel.py", line 1061, in __call__
    self.retrieve()
  File "/home/garar/.local/share/virtualenvs/pycaret-ml-W1zZIU60/lib/python3.8/site-packages/joblib/parallel.py", line 940, in retrieve
    self._output.extend(job.get(timeout=self.timeout))
  File "/home/garar/.local/share/virtualenvs/pycaret-ml-W1zZIU60/lib/python3.8/site-packages/joblib/_parallel_backends.py", line 542, in wrap_future_result
    return future.result(timeout=timeout)
  File "/usr/lib/python3.8/concurrent/futures/_base.py", line 439, in result
    return self.__get_result()
  File "/usr/lib/python3.8/concurrent/futures/_base.py", line 388, in __get_result
    raise self._exception
ValueError: y_true and y_pred contain different number of classes 3, 2. Please provide the true labels explicitly through the labels argument. Classes found in y_true: [0 1 2]
```

I don't think there is anything particularly wrong with logloss or my model, just that some combination causes, this code from `tabular.py` to exceute:

```
        except:
                    logger.error(f"create_model() for {model} raised an exception:")
                    logger.error(traceback.format_exc())
                    continue
```

which prevents this from execution (the `continue` at the end):

```
master_display_ = master_display.drop(
            ["Object", "runtime"], axis=1, errors="ignore"
        ).style.set_precision(round)
```

and `master_display_` is never assigned.

TLDR: exception has to be thrown by create model in `tabular.compare_models` for this issue to occur.

**Comment by Yard1** [comment]
Thanks, I'll look into this.

**Comment by Yard1** [comment]
@faruktufekci I just pushed commit 4a19b8b51d1e58bf2f9c4133bee603dd5a04d4c2 that should fix this issue. Could you run
pip install --upgrade --user git+git://github.com/pycaret/pycaret.git to install PyCaret's development version and try again?

**Comment by faruktufekci** [comment]
I'll try soon, thanks.
Did you commit this on purpose btw?

```
    display_container.append(compare_models_.data)
    display_container.append(compare_modfels_.data)
```

**Comment by Yard1** [comment]
No, it was a mistake. I have fixed it in a subsequent commit.

**Comment by rohitnig** [comment]
Any updates on this?

**Comment by pycaret** [comment]
@rohitnig The fix is already there on `master` branch. Can you please install pycaret from the master branch and test?

**Comment by garar** [comment]
Hi @pycaret !

It seems that now I get different error:

```
Traceback (most recent call last):
  File "pycaret-classification.py", line 79, in <module>
    run()
  File "pycaret-classification.py", line 57, in run
    top_models = compare_models(n_select=5)
  File "/home/garar/.local/share/virtualenvs/pycaret-ml-W1zZIU60/src/pycaret/pycaret/classification.py", line 764, in compare_models
    return pycaret.internal.tabular.compare_models(
  File "/home/garar/.local/share/virtualenvs/pycaret-ml-W1zZIU60/src/pycaret/pycaret/internal/tabular.py", line 2256, in compare_models
    for index, row in enumerate(master_display.iterrows()):
AttributeError: 'NoneType' object has no attribute 'iterrows'
```

It happens when I use `add_metric('logloss', 'LogLoss', log_loss, greater_is_better=False)`. It works without `logloss` metric.

**Comment by garar** [comment]
Great news! :)

I was able to reproduce using master and iris example:

```
from pycaret.classification import setup, compare_models, add_metric
from pycaret.datasets import get_data
from sklearn.metrics import log_loss

dataset = get_data('iris')

data = dataset.sample(frac=0.9, random_state=786)
data_unseen = dataset.drop(data.index)

data.reset_index(drop=True, inplace=True)
data_unseen.reset_index(drop=True, inplace=True)

print('Data for Modeling: ' + str(data.shape))
print('Unseen Data For Predictions: ' + str(data_unseen.shape))

exp_mclf101 = setup(data=data, target='species', session_id=123,
                    silent=True, html=False)

add_metric('logloss', 'LogLoss', log_loss, greater_is_better=False)

top_models = compare_models(n_select=5)
```

Traceback:

```
Traceback (most recent call last):                                                                                                                                                                                       
  File "iris.py", line 21, in <module>                                                       
    top_models = compare_models(n_select=5)                                                                                                                                                                                   
  File "/home/garar/.local/share/virtualenvs/pycaret-ml-W1zZIU60/src/pycaret/pycaret/classification.py", line 764, in compare_models
    return pycaret.internal.tabular.compare_models(                                                                                                                                                                       
  File "/home/garar/.local/share/virtualenvs/pycaret-ml-W1zZIU60/src/pycaret/pycaret/internal/tabular.py", line 2256, in compare_models
    for index, row in enumerate(master_display.iterrows()):                                                                                                                                                                    
AttributeError: 'NoneType' object has no attribute 'iterrows'  
```

PS. Sorry for the spam.

**Comment by Yard1** [comment]
Well, the issue here is twofold. The log loss metric seems to not be playing nicely with a multi label problem - but this is not something caused by PyCaret, as all we do is just pass this metric to sklearn and get the result. The other issue here are the exceptions that hapoen

**Comment by garar** [comment]
Interesting, sklearn docs don't mention that logloss is dedicated to binary classification: https://scikit-learn.org/stable/modules/model_evaluation.html#log-loss

I will try most recent version tomorrow and see if the issue when all models fail is fixed.

Thanks!

**Comment by Yard1** [comment]
The documentation says that:
```
labels array-like, optional (default=None)
If not provided, labels will be inferred from y_true. If labels is None and y_pred has shape (n_samples,) the labels are assumed to be binary and are inferred from y_true.
```

You can pass `labels` in `add_metric()` as a keyword argument. Perhaps that will help? You can also get the preprocessed target dataframe which this score function will be ran on with `get_config('y_test')`, so perhaps something like `np.unique(get_config('y_test'))` would give you an array you can use for `labels`.

**Comment by garar** [comment]
Hi @Yard1 !

Thanks for the tip about labels and thank you for the `master_display_` fix. It seems to be working. I've tested it using master branch.

Unless @faruktufekci objects, it seems like this can be closed.

**Comment by faruktufekci** [comment]
working on my machine as well. thank you @Yard1 !

**Comment by wsn4iv** [comment]
I receive the same error, but I am not adding any variables. I am just running compare_models. 

---------------------------------------------------------------------------
UnboundLocalError                         Traceback (most recent call last)
<ipython-input-19-cae85f799be1> in <module>()
----> 1 best_model=compare_models(exclude=['knn'])

~\AppData\Roaming\Python\Python37\site-packages\pycaret\classification.py in compare_models(include, exclude, fold, round, cross_validation, sort, n_select, budget_time, turbo, errors, fit_kwargs, groups, verbose)
    775         fit_kwargs=fit_kwargs,
    776         groups=groups,
--> 777         verbose=verbose,
    778     )
    779 

~\AppData\Roaming\Python\Python37\site-packages\pycaret\internal\tabular.py in compare_models(include, exclude, fold, round, cross_validation, sort, n_select, budget_time, turbo, errors, fit_kwargs, groups, verbose, display)
   2219     display.update_monitor(1, "Compiling Final Models")
   2220     display.display_monitor()
-> 2221 
   2222     display.move_progress()
   2223 

UnboundLocalError: local variable 'master_display_' referenced before assignment

**Comment by garar** [comment]
Hi @wsn4iv !

Does it happen on master? This error is indirectly caused by custom metrics, the main cause is: when pycaret can't create any models in compare_models it will fail, because `master_display_` is not initialized.

Did you try master?

**Comment by wsn4iv** [comment]
How do I try master?

Sent from my iPhone

Scott Newbern

**Comment by Yard1** [comment]
@wsn4iv `pip install --upgrade --user git+git://github.com/pycaret/pycaret.git` will install PyCaret's development version, which includes the fix for this issue.

**Comment by alanceloth** [comment]
Same error here:

```
UnboundLocalError: local variable 'master_display_' referenced before assignment
---------------------------------------------------------------------------
UnboundLocalError                         Traceback (most recent call last)
<ipython-input-336-e6b6f085ade8> in <module>
----> 1 best = compare_models(fold = 5)

D:\Anaconda\lib\site-packages\pycaret\regression.py in compare_models(include, exclude, fold, round, cross_validation, sort, n_select, budget_time, turbo, errors, fit_kwargs, groups, verbose)
    754     """
    755 
--> 756     return pycaret.internal.tabular.compare_models(
    757         include=include,
    758         exclude=exclude,

D:\Anaconda\lib\site-packages\pycaret\internal\tabular.py in compare_models(include, exclude, fold, round, cross_validation, sort, n_select, budget_time, turbo, errors, fit_kwargs, groups, verbose, display)
   2219 
   2220     compare_models_ = (
-> 2221         master_display_.apply(
   2222             highlight_max,
   2223             subset=[

UnboundLocalError: local variable 'master_display_' referenced before assignment
```

I will attach my notebook and log below:

The project is in a .rar folder in my onedrive, I can't upload directly here in github.
https://1drv.ms/u/s!Al5l7UUlb_GrhohIF_CSXycWmqmN-w?e=w8avzY

[logs.log](https://github.com/pycaret/pycaret/files/6713213/logs.log)
